cpp/src/barretenberg/bb/main.cpp

```diff
@@ -0,0 +1,183 @@
+    auto vk_data = from_buffer<plonk::verification_key_data>(read_file(vk_path));
+    acir_composer->load_verification_key(barretenberg::srs::get_crs_factory(), std::move(vk_data));
+    auto data = acir_composer->serialize_verification_key_into_fields();
+
+    // We need to move vk_hash to the front...
+    std::rotate(data.begin(), data.end() - 1, data.end());
+
+    auto json = format("[", join(map(data, [](auto fr) { return format("\"", fr, "\""); })), "]");
+    write_file(output_path, { json.begin(), json.end() });
+    info("vk as fields written to: ", output_path);
+}
+
+bool flagPresent(std::vector<std::string>& args, const std::string& flag)
+{
+    return std::find(args.begin(), args.end(), flag) != args.end();
+}
+
+std::string getOption(std::vector<std::string>& args, const std::string& option, const std::string& defaultValue)
+{
+    auto itr = std::find(args.begin(), args.end(), option);
+    return (itr != args.end() && std::next(itr) != args.end()) ? *(std::next(itr)) : defaultValue;
+}
+
+int main(int argc, char* argv[])
+{
+    std::vector<std::string> args(argv + 1, argv + argc);
+    verbose = flagPresent(args, "-v") || flagPresent(args, "--verbose");
+
+    if (args.empty()) {
+        std::cerr << "No command provided.\n";
+        return 1;
+    }
+
+    std::string command = args[0];
+
+    std::string json_path = getOption(args, "-j", "./target/main.json");
+    std::string witness_path = getOption(args, "-w", "./target/witness.tr");
+    std::string proof_path = getOption(args, "-p", "./proofs/proof");
+    std::string vk_path = getOption(args, "-k", "./target/vk");
+    bool recursive = flagPresent(args, "-r") || flagPresent(args, "--recursive");
+    init(json_path);
+
+    if (command == "prove_and_verify") {
+        proveAndVerify(json_path, witness_path, recursive);
+    } else if (command == "prove") {
+        std::string output_path = getOption(args, "-o", "./proofs/proof");
+        prove(json_path, witness_path, recursive, output_path);
+    } else if (command == "gates") {
+        gateCount(json_path);
+    } else if (command == "verify") {
+        verify(proof_path, recursive, vk_path);
+    } else if (command == "contract") {
+        std::string output_path = getOption(args, "-o", "./target/contract.sol");
+        contract(output_path, vk_path);
+    } else if (command == "write_vk") {
+        std::string output_path = getOption(args, "-o", "./target/vk");
+        writeVk(json_path, output_path);
+    } else if (command == "proof_as_fields") {
+        std::string output_path = getOption(args, "-o", proof_path + "_fields.json");
+        proofAsFields(proof_path, vk_path, output_path);
+    } else if (command == "vk_as_fields") {
+        std::string output_path = getOption(args, "-o", vk_path + "_fields.json");
+        vkAsFields(vk_path, output_path);
+    } else {
+        std::cerr << "Unknown command: " << command << "\n";
+        return -1;
+    }
```

Comment:
these are essentially public API's wonder if they could get abstracted away so they would be callable from wasm ...

Reply:
Not sure what you mean by callable from wasm. This binary is not meant to be used in/from any wasm context?